apps/docs/scripts/generate-embeddings.ts

```diff
@@ -0,0 +1,335 @@
+      metaExportNode.data.estree.body[0].declaration?.type === 'VariableDeclaration' &&
+      metaExportNode.data.estree.body[0].declaration.declarations[0]?.id.type === 'Identifier' &&
+      metaExportNode.data.estree.body[0].declaration.declarations[0].id.name === 'meta' &&
+      metaExportNode.data.estree.body[0].declaration.declarations[0].init?.type ===
+        'ObjectExpression' &&
+      metaExportNode.data.estree.body[0].declaration.declarations[0].init) ||
+    undefined
+
+  if (!objectExpression) {
+    return undefined
+  }
+
+  return getObjectFromExpression(objectExpression)
+}
+
+/**
+ * Splits a `mdast` tree into multiple trees based on
+ * a predicate function. Will include the splitting node
+ * at the beginning of each tree.
+ *
+ * Useful to split a markdown file into smaller sections.
+ */
+function splitTreeBy(tree: Root, predicate: (node: Content) => boolean) {
+  return tree.children.reduce<Root[]>((trees, node) => {
+    const [lastTree] = trees.slice(-1)
+
+    if (!lastTree || predicate(node)) {
+      const tree: Root = u('root', [node])
+      return trees.concat(tree)
+    }
+
+    lastTree.children.push(node)
+    return trees
+  }, [])
+}
+
+type Meta = ReturnType<typeof extractMetaExport>
+
+type ProcessedMdx = {
+  checksum: string
+  meta: Meta
+  sections: string[]
+}
+
+/**
+ * Processes MDX content for search indexing.
+ * It extracts metadata, strips it of all JSX,
+ * and splits it into sub-sections based on criteria.
+ */
+function processMdxForSearch(content: string): ProcessedMdx {
+  const checksum = createHash('sha256').update(content).digest('base64')
+
+  const mdxTree = fromMarkdown(content, {
+    extensions: [mdxjs()],
+    mdastExtensions: [mdxFromMarkdown()],
+  })
+
+  const meta = extractMetaExport(mdxTree)
+
+  // Remove all MDX elements from markdown
+  const mdTree = filter(
+    mdxTree,
+    (node) =>
+      ![
+        'mdxjsEsm',
+        'mdxJsxFlowElement',
+        'mdxJsxTextElement',
+        'mdxFlowExpression',
+        'mdxTextExpression',
+      ].includes(node.type)
+  )
+
+  if (!mdTree) {
+    return {
+      checksum,
+      meta,
+      sections: [],
+    }
+  }
+
+  const sectionTrees = splitTreeBy(mdTree, (node) => node.type === 'heading')
+
+  const sections = sectionTrees.map((tree) => toMarkdown(tree))
+
+  return {
+    checksum,
+    meta,
+    sections,
+  }
+}
+
+async function walk(dir: string): Promise<string[]> {
+  const immediateFiles = await readdir(dir)
+
+  const recursiveFiles = await Promise.all(
+    immediateFiles.map(async (file) => {
+      const filePath = join(dir, file)
+      const stats = await stat(filePath)
+      if (stats.isDirectory()) {
+        return walk(filePath)
+      } else if (stats.isFile()) {
+        return [filePath]
+      } else {
+        return []
+      }
+    })
+  )
+
+  const flattenedFiles = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  )
+
+  return flattenedFiles
+}
```

Comment:
This function is currently duplicated from `build-search.ts` in this POC with some minor type changes. Ideally we would re-use the same function for both.